noir-projects/noir-contracts/contracts/app/orderbook_contract/src/main.nr

```diff
@@ -0,0 +1,200 @@
+        } else {
+            (config.token1, config.token0)
+        };
+
+        // TODO(#14362): Once we allow partial note completion in private we can skip the transfer to public here and
+        // just finalize the maker's partial note straight away here.
+        // Transfer the amount_out from taker to the contract
+        Token::at(token_out)
+            .transfer_to_public(taker, context.this_address(), order.amount_out, nonce)
+            .call(&mut context);
+
+        // Prepare partial note for taker to receive token_in
+        let taker_partial_note =
+            Token::at(token_in).prepare_private_balance_increase(taker, taker).call(&mut context);
+
+        // Nullify the order such that it cannot be fulfilled again. We emit a nullifier instead of deleting the order
+        // from public storage because we get no refund for resetting public storage to zero and just emitting
+        // a nullifier is cheaper (1 Field in DA instead of multiple Fields for the order). We use the `order_id`
+        // itself as the nullifier because this contract does not work with notes and hence there is no risk of
+        // colliding with a real note nullifier.
+        //
+        // It's worth noting that since nullifier information is public, a malicious actor could potentially frontrun
+        // the taker's transaction. While this vulnerability exists in this simplified demonstration, a production
+        // implementation would typically employ a privileged matchmaker service to securely pair makers and takers,
+        // thereby ensuring order execution integrity.
+        context.push_nullifier(order_id.to_field());
+
+        // Enqueue the fulfillment to finalize both partial notes
+        Orderbook::at(context.this_address())
+            ._fulfill_order(
+                order_id,
+                taker_partial_note,
+                token_in,
+                token_out,
+                order.amount_in,
+                order.amount_out,
+            )
+            .enqueue(&mut context);
+    }
+
+    #[public]
+    #[internal]
+    fn _fulfill_order(
+        order_id: PartialUintNote, // maker partial note is used as `order_id`
+        taker_partial_note: PartialUintNote,
+        token_in: AztecAddress,
+        token_out: AztecAddress,
+        amount_in: u128,
+        amount_out: u128,
+    ) {
+        // Finalize transfer of amount_out of token_out to maker
+        Token::at(token_out).finalize_transfer_to_private(amount_out, order_id).call(&mut context);
+
+        // Finalize transfer of of amount_in of token_in to taker
+        Token::at(token_in).finalize_transfer_to_private(amount_in, taker_partial_note).call(
+            &mut context,
+        );
+
+        OrderFulfilled { order_id }.emit(encode_event(&mut context));
+    }
+
+    #[utility]
+    unconstrained fn get_order(order_id: PartialUintNote) -> pub Order {
+        storage.orders.at(order_id).read()
+    }
```

Comment:
Would be nice to also return a boolean indicating if the order was completed (by reading the nullifier)

Reply:
Good idea. Addressed in https://github.com/AztecProtocol/aztec-packages/pull/14464/commits/cf6973b3f43fda5484f472e4c8e857cbc379e3ff